### PR TITLE
Feature: III-3211

### PR DESCRIPTION
**Pull request by konrad126**
## Added
- the logic for logout action

[Jira](https://jira.uitdatabank.be/browse/III-3211)

### CODE REVIEW

File: app/RoutingServiceProvider.php [hunk 1]
```diff
@@ -26,15 +27,12 @@ function () {
                 $router->get('/connect', [RequestToken::class, '__invoke']);
                 $router->get('/authorize', [Authorize::class, '__invoke']);
 
-                // @TODO: implement
-                // $router->get('/register', [OAuthController::class, 'register']);
```
Comment (bertramakers) [inline app/RoutingServiceProvider.php]: Did you check with @Stiksels (Stan) what should happen to the `/register` endpoint?

File: src/Domain/Service/AuthServiceInterface.php [hunk 2]
```diff
@@ -2,11 +2,18 @@
...
 {
     public function redirectToLogin(): ?ResponseInterface;
 
+    /**
+     * @return string|null
+     * @throws UnSuccessfulAuthException
+     */
     public function token(): ?string;
+
+    public function logout(): void;
```
Comment (bertramakers) [inline src/Domain/Service/AuthServiceInterface.php]: Can we also add a `?ResponseInterface` return type here like for `redirectToLogin`? And maybe call it `redirectToLogout`?

File: tests/Domain/Action/LogoutTest.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    public function it_logs_out_user()
+    {
+        $serverRequest = $this->prophesize(ServerRequestInterface::class);
+        $destinationUrl = (new UriFactory())->createUri('http://foo-bar.com/');
+
+        $extractDestinationUrlFromRequest = $this->prophesize(ExtractDestinationUrlFromRequest::class);
+        $extractDestinationUrlFromRequest->__invoke($serverRequest)->willReturn($destinationUrl);
+
+        $authService = $this->prophesize(AuthServiceInterface::class);
```
Comment (bertramakers) [inline tests/Domain/Action/LogoutTest.php]: What's the reason you're using `$this->prophesize()` instead of `$this->createMock()` as we normally do in unit tests?
The `->reveal()` seems weird to me, and it's not required if you use `$this->createMock()`.

File: web/index.php [hunk 4]
```diff
@@ -12,8 +12,13 @@
 
 $container = ContainerFactory::forWeb($config);
 
-$response = $container->get(Router::class)->dispatch(
-    ServerRequestFactory::createFromGlobals()
-);
+try{
```
Comment (bertramakers) [inline web/index.php]: Nit: There should be a space between the `try` and `{` here, not sure why the coding standards don't complain about that.

File: web/index.php [hunk 5]
```diff
@@ -12,8 +12,13 @@
...
+    $response = $container->get(Router::class)->dispatch(
+        ServerRequestFactory::createFromGlobals()
+    );
+}catch(Throwable $throwable){
```
Comment (bertramakers) [inline web/index.php]: Same comment about the spaces as above.

File: src/Domain/Action/Logout.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        $this->extractDestinationUrlFromRequest = $extractDestinationUrlFromRequest;
+        $this->authService = $authService;
+        $this->responseFactory = $responseFactory;
+    }
+
+    public function __invoke(ServerRequestInterface $serverRequest): ResponseInterface
+    {
+        $destinationUrl = $this->extractDestinationUrlFromRequest->__invoke($serverRequest);
+        $this->authService->logout();
+        return $this->responseFactory->redirectTo($destinationUrl);
```
Comment (bertramakers) [inline src/Domain/Action/Logout.php]: I'm a bit confused here. Doesn't `$this->authService->logout()` redirect to Auth0, and then Auth0 is supposed to redirect to a whitelisted logout callback and only then do we redirect back to the destination?

Comment (bertramakers) [reply src/Domain/Action/Logout.php]: In the Auth0 PHP docs it says we have to redirect to Auth0 after calling `logout()`
https://auth0.com/docs/quickstart/webapp/php/#logout

Comment (bertramakers) [reply src/Domain/Action/Logout.php]: See also https://auth0.com/docs/libraries/auth0-php/authentication-api#single-sign-on-logout

File: tests/Domain/Action/LogoutTest.php (hunk 3 above)
Comment (konrad126) [reply tests/Domain/Action/LogoutTest.php]: I like prophecy better than the Mockery (I think this is used when you do `$this->getMock()`), I think it's more expressive but it is a bit annoying you have to do the `reveal()`, in `phpspec` this is done in "the background" so you don't have to do it explicitly. 

Creator of `PHPUnit` also thinks prophecy it better :) : https://thephp.cc/news/2015/02/phpunit-4-5-and-prophecy

Comment (bertramakers) [reply tests/Domain/Action/LogoutTest.php]: Ah okay. What I like about PHPUnit's mocking library is that you get easy auto-completion on for example `$this->createMock(...)->expects($this->once())->with(...)->willReturn(...)` etc.

If you look at an example from the blog post:
```
$reader = $this->prophesize('Reader') ;
$reader->read( '123' )->willReturn('value')->shouldBeCalled();
```

I'm pretty sure PHPStorm won't be able to auto-complete the `willReturn` (or jump to definition etc) because it doesn't know what the magic `read` method will return?

I value easy autocompletion and jump-to-definition because it makes debugging / changing code in the future a lot easier.

Comment (bertramakers) [reply tests/Domain/Action/LogoutTest.php]: But I'm not entirely opposed to use prophecy if you like it better, I'm just not entirely convinced either 😄 

We can merge it like this though. When the other comments get resolved of course.

Comment (konrad126) [reply tests/Domain/Action/LogoutTest.php]: With this plugin `https://plugins.jetbrains.com/plugin/9674-phpunit-enhancement/versions` PHP Storm recognizes it :)

File: app/RoutingServiceProvider.php (hunk 1 above)
Comment (bertramakers) [reply app/RoutingServiceProvider.php]: @konrad126 can you create a follow-up ticket in Jira to remove the register link from the front-end application so we don't forget? Thanks!

File: app/RoutingServiceProvider.php [hunk 7]
```diff
@@ -26,15 +28,14 @@ function () {
...
+                $router->get('/logout', [RequestLogout::class, '__invoke']);
+                $router->get('/logout-confirm', [LogOut::class, '__invoke']);
 
                 // Maintain these old paths for backwards compatibility.
                 $router->get('/culturefeed/oauth/connect', [RequestToken::class, '__invoke']);
                 $router->get('/culturefeed/oauth/authorize', [Authorize::class, '__invoke']);
-                // $router->get('/culturefeed/oauth/register', [OAuthController::class, 'register']);
-                // $router->get('/culturefeed/oauth/logout', [OAuthController::class, 'logout']);
+                $router->get('/culturefeed/oauth/logout', [RequestLogout::class, '__invoke']);
+                $router->get('/culturefeed/oauth/logout-confirm', [LogOut::class, '__invoke']);
```
Comment (bertramakers) [inline app/RoutingServiceProvider.php]: Nit: `/culturefeed/oauth/logout-confirm'` didn't exist before so we don't need to keep it for bc.
But I think it'll be removed again anyway in another PR?

File: config.dist.yml [hunk 8]
```diff
@@ -18,3 +18,4 @@ auth0:
   client_id:
   client_secret:
   redirect_uri:
+  log_out_uri:
```
Comment (bertramakers) [inline config.dist.yml]: Remember to create an `appconfig` PR for this!

File: web/index.php [hunk 9]
```diff
@@ -16,4 +16,5 @@
     ServerRequestFactory::createFromGlobals()
 );
 
+
```
Comment (bertramakers) [inline web/index.php]: Nit: This newline seems unnecessary?

File: app/RoutingServiceProvider.php (hunk 7 above)
Comment (konrad126) [reply app/RoutingServiceProvider.php]: Yes, we'll remove it in the next one :)

**Comment by konrad126** [comment]
@bertramakers yeah the service was starting to get bloated.. :)
